### PR TITLE
feat: claims

### PR DESCRIPTION
**Pull request by srindom**
Adds support for claims handling in Medusa.

**Behavior**

- Can behave in different ways:
    - Order was lost → Send replacement (`type: "replace"`, `item.reason: "missing_item"`)
    - Product has production failure → Record why, send replacement (`type: "replace"`)
        - Optionally send a return label (*select return address? → requires address settings)*
    - Customer prefers a refund → refund product amount (`type: "refund"`)
    - Order has wrong items → Recieve the wrong items and send new (`type: "replace"`, `item.reason: "wrong_item"`)

### CODE REVIEW

File: packages/medusa/src/services/claim.js [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+  update(id, data) {
+    return this.atomicPhase_(async manager => {
+      const claimRepo = manager.getCustomRepository(this.claimRepository_)
+      const claim = await this.retrieve(id, { relations: ["shipping_methods"] })
+
+      const { shipping_methods, metadata } = data
+
+      if (metadata) {
+        claim.metadata = this.setMetadata_(claim, update.metadata)
+        claimRepo.save(claim)
```
Comment (olivermrbl) [inline packages/medusa/src/services/claim.js:103]: await

File: packages/medusa/src/services/claim.js [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+      await Promise.all(
+        claim.fulfillments.map(f =>
+          this.fulfillmentService_.withTransaction(manager).cancelFulfillment(f)
+        )
+      )
+
+      claim.fulfillment_status = "canceled"
+
+      const claimRepo = manager.getCustomRepository(this.claimRepository_)
+      const result = claimRepo.save(claim)
```
Comment (olivermrbl) [inline packages/medusa/src/services/claim.js:493]: await

File: packages/medusa/src/services/claim.js [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+  /**
+   * Dedicated method to set metadata for an order.
+   * To ensure that plugins does not overwrite each
+   * others metadata fields, setMetadata is provided.
+   * @param {string} orderId - the order to decorate.
+   * @param {string} key - key for metadata field
+   * @param {string} value - value for metadata field.
+   * @return {Promise} resolves to the updated result.
+   */
+  setMetadata_(order, metadata) {
```
Comment (olivermrbl) [inline packages/medusa/src/services/claim.js:549]: Remove as this is part of the base service

**Review by olivermrbl — CHANGES_REQUESTED** [review]
This is an amazing feature. Good job man!
Minor, minor comments...

File: packages/medusa/src/migrations/1612170081396-claims_1.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import { MigrationInterface, QueryRunner } from "typeorm"
```
[the remark this answers is not in the document]
Comment (srindom) [reply packages/medusa/src/migrations/1612170081396-claims_1.ts:1]: agree

File: packages/medusa/src/services/claim.js (hunk 1 above)
Comment (srindom) [reply packages/medusa/src/services/claim.js:103]: good catch

**Pull request merged**